path_selection/path_selection.go

```
@@ -0,0 +1,275 @@
+func getDelays(lambda float64, count int) []float64 {
+	cryptRand := rand.NewMath()
+	delays := make([]float64, count)
+	for i := 0; i < count-2; i++ {
+		delays[i] = rand.Exp(cryptRand, lambda)
+	}
+	delays[count-1] = float64(0)
+	return delays
+}
+
+// sum adds a slice of float64.
+// this is used to get the sum of delays
+// which are represented as float64s
+func sum(input []float64) float64 {
+	total := float64(0)
+	for _, n := range input {
+		total += n
+	}
+	return total
+}
+
+// RouteFactory builds routes and by doing so handles all the
+// details of path selection and mix PKI interaction. This factory
+// is used to create routes for a mix network which uses the
+// Poisson mix strategy, where the client selects a delay for each
+// hop in the route where the mean of this distribution is tuneable
+// using the lambda parameter.
+type RouteFactory struct {
+	pki    pki.Client
+	nrHops int
+	lambda float64
+}
+
+// NewRouteFactory creates a new RouteFactory for creating routes
+func New(pki pki.Client, nrHops int, lambda float64) *RouteFactory {
+	r := RouteFactory{
+		pki:    pki,
+		nrHops: nrHops,
+		lambda: lambda,
+	}
+	return &r
+}
+
+// getRouteDescriptors returns a slice of mix descriptors,
+// one for each hop in the route where each mix descriptor
+// was selected from the set of descriptors for that layer
+func (r *RouteFactory) getRouteDescriptors(senderProviderName, recipientProviderName string) ([]*pki.MixDescriptor, error) {
+	var err error
+	descriptors := make([]*pki.MixDescriptor, r.nrHops)
+	descriptors[0], err = r.pki.GetProviderDescriptor(senderProviderName)
+	if err != nil {
+		return nil, err
+	}
+	descriptors[r.nrHops-1], err = r.pki.GetProviderDescriptor(recipientProviderName)
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < r.nrHops; i++ {
+		layerMixes := r.pki.GetMixesInLayer(uint8(i))
+		fmt.Printf("layer %d\n", i)
+		if len(layerMixes) == 0 {
+			panic("WTF")
```

why not return an error? if this does warrant panic, then provide a useful error

---

ok... i'll fix this